cpp/src/arrow/builder.cc

```diff
@@ -121,6 +121,14 @@ void ArrayBuilder::UnsafeAppendToBitmap(const uint8_t* valid_bytes, int64_t leng
   uint8_t bitset = null_bitmap_data_[byte_offset];
 
   for (int64_t i = 0; i < length; ++i) {
+    if (bit_offset == 8) {
```

Comment:
When the length is a power of 2, on the last iteration of the loop, a value past the end of the null bitmap is accessed, causing valgrind to complain. I only stumbled on this by accident by using 4096 as the length in my test case